Iceberg support for reading v2 row level deletes

alexjo2144 opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->

## Description

Iceberg support for reading Parquet row level deletes.

Depends on https://github.com/trinodb/docker-images/pull/123 for testing.

Based off of https://github.com/trinodb/trino/pull/10075

<!-- Elaborate beyond the title of the PR as necessary to help the reviewers and maintainers.-->

<!-- Answer the following questions to help reviewers and maintainers
understand this PR's scope at a glance:
-->

> Is this change a fix, improvement, new feature, refactoring, or other?

New feature

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

Connector, Iceberg

> How would you describe this change to a non-technical end user or system administrator?

Support for v2 merge-on-read reads from Parquet tables.

## Related issues, pull requests, and links

<!-- List any issues fixed by this PR, and provide links to other related PRs, upstream release notes, and other useful resources. For example:
* Fixes #issuenumber
* Related documentation in #issuenumber
* [Some release notes](http://usefulinfo.example.com)
-->

<!-- The following sections are filled in by the maintainer with input from the contributor:
Use :white_check_mark: or (x) to signal selection.
-->

## Documentation

( ) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
( ) Release notes entries required with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

alexjo2144 commented:
Marking this a WIP since there's no testing yet

alexjo2144 commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/delete/DummyFileScanTask.java:
```diff
@@ -0,0 +1,230 @@
+
+import static java.util.Objects.requireNonNull;
+
+// TODO: This wrapper is necessary until the constructors of the Iceberg DeleteFilter are made more specific
```
https://github.com/apache/iceberg/pull/4381

samarthjain commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSource.java:
```diff
@@ -38,35 +44,58 @@
+        }
+
+        // allColumns should include all expectedColumns as well as any columns needed by the DeleteFilter
+        requireNonNull(allColumns, "allColumns is null");
```
This check should be right after 
```
 requireNonNull(expectedColumns, "expectedColumns is null");
 ```

samarthjain commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:
```diff
@@ -184,17 +194,35 @@ public ConnectorPageSource createPageSource(
+                .forEach(columnsForDataPageSource::add);
+        List<IcebergColumnHandle> deleteFilterRequirements = getColumns(deleteFilter.requiredSchema(), typeManager);
+        deleteFilterRequirements.stream()
+                .filter(deleteRequiredColumn -> !icebergColumns.contains(deleteRequiredColumn))
```
Since we are doing a distinct here:
```
 List<IcebergColumnHandle> allColumns = Streams.concat(icebergColumns.stream(), deleteFilterRequirements.stream())
                .distinct()
                .collect(toImmutableList());
```, 
we can remove this filter on the stream.

alexjo2144 commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/delete/TrinoDeleteFilter.java:
```diff
@@ -0,0 +1,50 @@
+
+    public TrinoDeleteFilter(FileScanTask task, Schema tableSchema, List<IcebergColumnHandle> requestedColumns, FileIO fileIO)
+    {
+        super(task, tableSchema, schemaForColumnHandles(requestedColumns));
```
Turns out this util method does not set field ids properly, will need to try something else.

findepi commented on plugin/trino-iceberg/pom.xml:
```diff
@@ -169,6 +169,22 @@
+                <exclusion>
+                    <groupId>org.apache.parquet</groupId>
+                    <artifactId>parquet-avro</artifactId>
+                </exclusion>
```
document why the exclusions

findepi commented on plugin/trino-iceberg/pom.xml:
```diff
@@ -169,6 +169,22 @@
+                </exclusion>
+                <exclusion>
+                    <groupId>org.apache.orc</groupId>
+                    <artifactId>orc-core</artifactId>
```
cc @homar . will this be a problem for ORC path?

findepi commented on plugin/trino-iceberg/pom.xml:
```diff
@@ -187,6 +203,11 @@
+        <dependency>
+            <groupId>org.jetbrains</groupId>
+            <artifactId>annotations</artifactId>
+        </dependency>
```
`<scope>provided</scope>`

alexjo2144 replied:
Maven complains that it is not provided

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergTableHandle.java:
```diff
@@ -59,7 +60,8 @@ public IcebergTableHandle(
             @JsonProperty("projectedColumns") Set<IcebergColumnHandle> projectedColumns,
-            @JsonProperty("nameMappingJson") Optional<String> nameMappingJson)
+            @JsonProperty("nameMappingJson") Optional<String> nameMappingJson,
+            @JsonProperty("schemaJson") String schemaJson)
```
This should come after snapshotId field / param

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/delete/DummyFileScanTask.java:
```diff
@@ -0,0 +1,231 @@
+import static java.util.Objects.requireNonNull;
+
+// TODO: This wrapper is necessary until the constructors of the Iceberg DeleteFilter are made more specific
+// Remove after merging https://github.com/apache/iceberg/pull/4381
```
```suggestion
// Remove after upgrading to iceberg with https://github.com/apache/iceberg/pull/4381
```

findepi replied:
Can we have eg reflection-based test to ensure the TODO gets done?

alexjo2144 replied:
This line will stop compiling after pulling in the updated Iceberg library with that PR https://github.com/trinodb/trino/pull/11642/files#diff-bea7e01e5f9c44ac82298bfe3334cc2db359a751723913112135494273025cd8R40

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/delete/DummyFileScanTask.java:
```diff
@@ -0,0 +1,231 @@
+    @Override
+    public PartitionSpec spec()
+    {
+        throw new IllegalArgumentException("Cannot call 'spec' method on DummyFileScanTask");
```
this would be sufficient:

```suggestion
        throw new UnsupportedOperationException();
```

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/delete/TrinoDeleteFile.java:
```diff
@@ -0,0 +1,317 @@
+public class TrinoDeleteFile
+        implements DeleteFile
+{
+    public static final long INSTANCE_SIZE = ClassLayout.parseClass(IcebergSplit.class).instanceSize();
```
wrong class

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplit.java:
```diff
@@ -128,7 +139,8 @@ public long getRetainedSizeInBytes()
                 + estimatedSizeOf(addresses, HostAddress::getRetainedSizeInBytes)
-                + estimatedSizeOf(partitionKeys, SizeOf::sizeOf, valueOptional -> sizeOf(valueOptional, SizeOf::estimatedSizeOf));
+                + estimatedSizeOf(partitionKeys, SizeOf::sizeOf, valueOptional -> sizeOf(valueOptional, SizeOf::estimatedSizeOf))
+                + TrinoDeleteFile.INSTANCE_SIZE * deletes.size(); // TODO: TrinoDeleteFile has some complex fields, this size will be an underestimate
```
address the TODO

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/delete/TrinoDeleteFile.java:
```diff
@@ -0,0 +1,317 @@
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+
+/**
+ * A Jackson Serializable implementation of the Iceberg DeleteFile. Allows for delete files to be included in IcebergSplit.
```
```suggestion
 * A Jackson-serializable implementation of the Iceberg {@link DeleteFile}. Allows for delete files to be included in {@link IcebergSplit}.
```

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/delete/TrinoDeleteFile.java:
```diff
@@ -0,0 +1,317 @@
+    private final Integer sortOrderId;
+    private final List<Long> splitOffsets;
+
+    public static TrinoDeleteFile fromIcebergDeleteFile(DeleteFile icebergDeleteFile)
```
`copyOf` ?

icebergDeleteFile -> deleteFile

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/delete/TrinoDeleteFile.java:
```diff
@@ -0,0 +1,317 @@
+            @JsonProperty("sortOrderId") Integer sortOrderId,
+            @JsonProperty("splitOffsets") List<Long> splitOffsets)
+    {
+        this.pos = pos;
```
add `rnn` where applicable, and `@Nullable` (on constructor param and getters) elsewhere.

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/delete/TrinoDeleteFile.java:
```diff
@@ -0,0 +1,317 @@
+        return format;
+    }
+
+    // TODO: Probably need to figure out how to serialize this
+    @Override
+    public StructLike partition()
+    {
+        throw new IllegalArgumentException("Cannot call 'partition' method on TrinoDeleteFile");
```
If this is not needed, kist add `@JsonIgnore` + `throw new UnsupportedOperationException()`

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/delete/TrinoDeleteFile.java:
```diff
@@ -0,0 +1,317 @@
+    }
+
+    @Override
+    public DeleteFile copy()
+    {
+        return new TrinoDeleteFile(
```
Can we return `this`?
I'd hope our class is immutable

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/delete/TrinoDeleteFilter.java:
```diff
@@ -0,0 +1,102 @@
+    public TrinoDeleteFilter(FileScanTask task, Schema tableSchema, List<IcebergColumnHandle> requestedColumns, FileIO fileIO)
+    {
+        super(task, tableSchema, filterSchema(tableSchema, requestedColumns));
+        this.fileIO = fileIO;
```
`rnn`

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/delete/TrinoDeleteFilter.java:
```diff
@@ -0,0 +1,102 @@
+            if (requiredChildren.isEmpty()) {
+                return Optional.empty();
+            }
+            else {
```
redundant else

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/delete/TrinoDeleteFilter.java:
```diff
@@ -0,0 +1,102 @@
+            }
+        }
+
+        else {
```
redundant else

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:
```diff
@@ -184,17 +194,35 @@ public ConnectorPageSource createPageSource(
+        FileIO fileIO = fileIoProvider.createFileIo(hdfsContext, session.getQueryId());
+        TrinoDeleteFilter deleteFilter = new TrinoDeleteFilter(
+                new DummyFileScanTask(split.getPath(), split.getDeletes()),
+                SchemaParser.fromJson(table.getSchemaJson()),
+                icebergColumns,
```
This ends up filtering table schema looking for icebergColumns (columns being read by the query).

- for positional deletes, this is useless, since read columns have nothing to do with positions;
- for equality-based deletes, this seems wrong, since equality delete can use a column that is **not** being read

For the `TrinoDeleteFilter` would it work if we pass `tableSchema` as `requestedSchema` to `DeleteFilter` constructor?
would `org.apache.iceberg.data.DeleteFilter#requiredSchema` still be correct?

alexjo2144 replied:
> for equality-based deletes, this seems wrong, since equality delete can use a column that is not being read

This is where it's used. The DeleteFilter uses this list to see what columns were going to be read anyway by the query, and adds any columns needed to apply the filter to the `DeleteFilter.requiredSchema` list.

For positional deletes, this `requiredSchema` also includes the row number column.

alexjo2144 replied:
> if we pass tableSchema as requestedSchema to DeleteFilter

This would result in all queries reading every column of the data file

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:
```diff
@@ -645,30 +684,43 @@ private static ReaderPageSource createParquetPageSource(
                     messageColumnIO,
                     blocks,
-                    Optional.empty(),
+                    Optional.of(blockStarts.build()),
```
Looks like in ParquetReader, the `firstRowsOfBlocks` should no longer be Optional (change as a cleanup commit, add after the main change, or separate pr)

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/delete/TrinoRow.java:
```diff
@@ -0,0 +1,94 @@
+        throw new TrinoException(NOT_SUPPORTED, "writing to TrinoRow is not supported");
+    }
+
+    public static Iterable<TrinoRow> fromBlocks(Type[] types, Block[] blocks, int batchSize)
```
batchSize must be equal the length of each block, so the argument name is wrong
(otherwise one'd expect a starting index)

call it `positionCount`

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/delete/TrinoRow.java:
```diff
@@ -0,0 +1,94 @@
+                    @Override
+                    public TrinoRow next()
+                    {
+                        return new TrinoRow(types, blocks, i++);
```
- avoid  prefix/postfix increment in expression context. treat it as a statement
- Iterator.next() must throw a specific ex if `hasNext()`  would return false

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/delete/TrinoRow.java:
```diff
@@ -0,0 +1,94 @@
+    {
+        return new Iterable<>()
+        {
+            private int i;
```
nextPosition

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSource.java:
```diff
@@ -38,35 +44,58 @@
     {
-        int size = requireNonNull(columns, "columns is null").size();
+        // expectedColumns should contain columns which should be in the final Page
+        // allColumns should include all expectedColumns as well as any columns needed by the DeleteFilter
```
since it's not "all columns of a table", don't call it allColumns. requiredColumns?

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSource.java:
```diff
@@ -38,35 +44,58 @@
+        requireNonNull(allColumns, "allColumns is null");
+        this.expectedColumnIndexes = new int[expectedColumns.size()];
+        for (int i = 0; i < expectedColumns.size(); i++) {
+            expectedColumnIndexes[i] = allColumns.indexOf(expectedColumns.get(i));
```
This is quadratic. 
Can we expect `expectedColumns` and `allColumns` be in same order?

alexjo2144 replied:
I need to rework this a little to handle projection pushdown, will see if I can make it linear too.

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSource.java:
```diff
@@ -120,7 +149,13 @@ public Page getNextPage()
-            return new Page(batchSize, blocks);
+
+            CloseableIterable<TrinoRow> filteredRows = deleteFilter.filter(CloseableIterable.withNoopClose(TrinoRow.fromBlocks(columnTypes, blocks, batchSize)));
+            int[] positionsToKeep = StreamSupport.stream(filteredRows.spliterator(), false)
```
`[Streams.]stream(filteredRows)`

however, we avoid using streams in "execution" code (eg invoked per row or per page), for their slight overhead.

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSource.java:
```diff
@@ -120,7 +149,13 @@ public Page getNextPage()
+            int[] positionsToKeep = StreamSupport.stream(filteredRows.spliterator(), false)
+                    .mapToInt(TrinoRow::getPosition)
+                    .toArray();
+            Page pageWithPrefilledValues = new Page(batchSize, blocks);
```
batchSize -> positionCount

findepi commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergV2.java:
```diff
@@ -1,209 +0,0 @@
-import static io.trino.testing.sql.TestTable.randomTableSuffix;
-import static io.trino.tpch.TpchTable.NATION;
-
-public class TestIcebergV2
```
why is this removed?

findepi reviewed (commented):
Can you please add some Spark-based PTs?

alexjo2144 committed:
Update testing Docker image tag to 55

alexjo2144 commented:
@homar I pulled in the parts of your ORC changes that were necessary for reads https://github.com/trinodb/trino/pull/11543

Also added some Spark compatibility tests with positional deletes. Still need to add some with equality based deletes.

findinpath commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergColumnHandle.java:
```diff
@@ -133,11 +135,25 @@ public String getQualifiedName()
+    }
+
+    @JsonIgnore
+    // Marker column used by the Iceberg DeleteFilter to indicate columns which are deleted by equality deletes
```
`to indicate columns` -> `to indicate rows` right?

alexjo2144 replied:
Absolutely. Thanks for catching

alexjo2144 committed:
Make ParquetReader firstRowsOfBlocks required

alexjo2144 commented:
@findepi I still need to add some equality based delete tests, and some with projection pushdown probably but mind taking another look?

The big change in the last push is that I changed the way partition values are filled in. The used to be done in the IcebergPageSource but it simplified things a lot to change the code to handle them in the ORC/Parquet readers, so there's a new ColumnAdaptation for ORC that fills in the constant partition value, and a similar change for Parquet.

groobyming commented:
@alexjo2144 
Hi, I merged the code of this pr and did a test locally, and found that the read speed of copy on read and merge on read are very different. For the same amount of data, merge on read is almost 800 times slower than copy on write. Below are my test results:

reading method: merge on read
![image](https://user-images.githubusercontent.com/34335395/160780910-73f51ade-2d41-4cd5-b10e-db25d89bf1fe.png)
cost time: 1635 seconds

reading method: copy on write
![image](https://user-images.githubusercontent.com/34335395/160781083-b513348b-b3ce-4b89-9695-bcf2dedf8955.png)
cost time: 2.02 seconds

samarthjain commented:
@groobyming - did you mean to say that query speeds are different when files didn't contain deleted rows because they were re-written using copy on write mechanism vs filtering out deleted rows during read time?

groobyming commented:
> 

es

groobyming commented:
@samarthjain 
Yeah, that's what I meant

samarthjain commented:
@groobyming How many rows were deleted cause that can have a significant impact on merge on read.

groobyming commented:
@samarthjain 
About 550 million

samarthjain commented:
@groobyming - that potentially could be a lot. In general, copy on write approach works best for reading data where as merge-on-read is more efficient at writing deletes.

groobyming commented:
@samarthjain Thanks for your reply, for best read performance maybe I should use copy-on-write when deleting data

samarthjain commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/ConstantPopulatingPageSource.java:
```diff
@@ -0,0 +1,180 @@
+        }
+
+        for (Map.Entry<Integer, Block> constantColumnEntry : constantColumns.entrySet()) {
+            blocks[constantColumnEntry.getKey()] = new RunLengthEncodedBlock(constantColumnEntry.getValue(), delegatePage.getPositionCount());
```
There is an inherent assumption here that `sourceChannelToTargetChannel.get(sourceChannel)` will always be different from `constantColumnEntry.getKey()` . Should we validate that in the constructor?

samarthjain commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:
```diff
@@ -645,47 +717,79 @@ private static ReaderPageSource createParquetPageSource(
-                trinoTypes.add(trinoType);
-
-                if (parquetField == null) {
+                if (column.isIcebergIsDeletedMetadataColumn()) {
```
This made me think whether we should instead have the method name simply as `isDeletedMetadataColumn()` since the column handle is already an `IcebergColumnHandle`.

samarthjain commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:
```diff
@@ -645,47 +717,79 @@ private static ReaderPageSource createParquetPageSource(
+                            trinoType,
+                            deserializePartitionValue(trinoType, partitionKeys.get(column.getId()).orElse(null), column.getName()))));
+                }
+                else if (column.isIcebergRowPositionMetadataColumn()) {
```
Similarly, `isRowPositionMetadataColumn()`

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/delete/TrinoDeleteFile.java:
```diff
@@ -0,0 +1,339 @@
+    private final FileFormat format;
+    private final long recordCount;
+    private final long fileSizeInBytes;
+    @Nullable private final Map<Integer, Long> columnSizes;
+    @Nullable private final Map<Integer, Long> valueCounts;
+    @Nullable private final Map<Integer, Long> nullValueCounts;
+    @Nullable private final Map<Integer, Long> nanValueCounts;
+    @Nullable private final Map<Integer, byte[]> lowerBounds;
+    @Nullable private final Map<Integer, byte[]> upperBounds;
```
would it make the code simpler if we had empty maps instead of nulls?

the `copyOf` would still need to handle the flexibility of the `DeleteFile` (having `firstNonNull(..., ImmutableMap.of())` would be sufficient for that, without `?:`),
and the rest of the class would be simpler

or, is it a semantical difference between an empty map and a null?

alexjo2144 replied:
>  is it a semantical difference between an empty map and a null?

There isn't in our code but the Iceberg API treats the Maps as nullable and I don't know if there's a distinction in how it's used there so I'd rather leave it.

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/delete/TrinoDeleteFile.java:
```diff
@@ -0,0 +1,339 @@
+            @JsonProperty("format") FileFormat format,
+            @JsonProperty("recordCount") long recordCount,
+            @JsonProperty("fileSizeInBytes") long fileSizeInBytes,
+            @JsonProperty("columnSizes") @org.jetbrains.annotations.Nullable Map<Integer, Long> columnSizes,
```
import Nullable statically

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/delete/TrinoDeleteFile.java:
```diff
@@ -0,0 +1,339 @@
+        this.nanValueCounts = nanValueCounts == null ? null : ImmutableMap.copyOf(nanValueCounts);
+        this.lowerBounds = lowerBounds == null ? null : ImmutableMap.copyOf(lowerBounds);
+        this.upperBounds = upperBounds == null ? null : ImmutableMap.copyOf(upperBounds);
+        this.keyMetadata = keyMetadata;
```
`keyMetadata` is mutable, `.clone()` it

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/delete/TrinoDeleteFile.java:
```diff
@@ -0,0 +1,339 @@
+            return null;
+        }
+        return upperBounds.entrySet().stream()
+                .collect(toImmutableMap(Map.Entry<Integer, byte[]>::getKey, entry -> ByteBuffer.wrap(entry.getValue())));
```
> `Map.Entry<Integer, byte[]>::getKey`

`Entry::getKey` would be sufficient

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/delete/TrinoDeleteFile.java:
```diff
@@ -0,0 +1,339 @@
+
+    public long getRetainedSizeInBytes()
+    {
+        ToLongFunction<Integer> intSizeOf = ignored -> Integer.SIZE / 8;
```
`SizeOf.SIZE_OF_INT`

findepi replied:
actually, `SizeOf.INTEGER_INSTANCE_SIZE`

alexjo2144 replied:
https://github.com/airlift/slice/pull/147

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/delete/TrinoDeleteFile.java:
```diff
@@ -0,0 +1,339 @@
+                + estimatedSizeOf(path)
+                + estimatedSizeOf(columnSizes, intSizeOf, longSizeOf)
+                + estimatedSizeOf(nullValueCounts, intSizeOf, longSizeOf)
+                + estimatedSizeOf(nanValueCounts, intSizeOf, longSizeOf)
```
We should add `estimatedSizeOf(Map<K, V> map, long keySize, long valueSize)` overload where key and value sizes are constant. This would avoid map traversal.

findepi replied:
You can leverage https://github.com/airlift/slice/pull/146 now.

findepi replied:
right

findepi replied:
let's apply https://github.com/trinodb/trino/pull/11642#discussion_r842859668 and release airbase too

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/delete/TrinoRow.java:
```diff
@@ -0,0 +1,101 @@
+    @Override
+    public <T> void set(int i, T t)
+    {
+        throw new TrinoException(NOT_SUPPORTED, "writing to TrinoRow is not supported");
```
UnsupportedOperationException

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/delete/TrinoRow.java:
```diff
@@ -0,0 +1,101 @@
+
+    public static Iterable<TrinoRow> fromPage(Type[] types, Page page, int positionCount)
+    {
+        return new Iterable<>()
+        {
+            private int nextPosition;
+
+            @Override
+            public Iterator<TrinoRow> iterator()
+            {
+                return new Iterator<>() {
+                    @Override
+                    public boolean hasNext()
+                    {
+                        return nextPosition < positionCount;
+                    }
+
+                    @Override
+                    public TrinoRow next()
+                    {
+                        if (!hasNext()) {
+                            throw new NoSuchElementException();
+                        }
+
+                        TrinoRow next = new TrinoRow(types, page, nextPosition);
+                        nextPosition++;
+                        return next;
+                    }
+                };
+            }
+        };
```
```java
        return () -> new AbstractIterator<>()
        {
            private int nextPosition;

            @Override
            protected TrinoRow computeNext()
            {
                if (nextPosition == positionCount) {
                    return endOfData();
                }
                int position = nextPosition;
                nextPosition++;
                return new TrinoRow(types, page, position);
            }
        };
```

findepi commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergV2.java:
```diff
@@ -145,20 +145,8 @@ public void testV2TableWithEqualityDelete()
-    }
-
-    @Test
-    public void testV2TableWithUnreadEqualityDelete()
```
why removed?

alexjo2144 replied:
It was there to make sure that delete files only caused a failure if the query actually read files that needed them, but now the query should always work so it's redundant with the test above it

alexjo2144 commented on plugin/trino-iceberg/pom.xml:
```diff
@@ -353,6 +389,17 @@
                 </configuration>
             </plugin>
+
+            <!-- iceberg-orc is a provided dependency of iceberg-data -->
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-dependency-plugin</artifactId>
+                <configuration>
+                    <usedDependencies>
+                        <dependency>org.apache.iceberg:iceberg-orc</dependency>
```
I'll update the comment with more info but `iceberg-orc` needs to be supplied as a provided dependency of `iceberg-data`, however we're not using any code directly from that library so if I add it as a dependency the Maven Enforcer rule fails with a "Unused declared dependencies found" error.

Explicitly marking it as "used" here gets rid of that error, but maybe there's a better way?

findepi replied:
provide dependency as `runtime` scope?

alexjo2144 replied:
Aha, thanks I knew I was missing something.

alexjo2144 commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:
```diff
@@ -388,14 +457,16 @@ private static ReaderPageSource createOrcPageSource(
-                            Optional.empty(),
-                            memoryUsage,
-                            stats),
+                    new ConstantPopulatingPageSource(
+                            new OrcPageSource(
```
I'll try to remove it for queries where it's not necessary. As for the ColumnAdaptations, that's how I had it originally but it required changes to both ORC and Parquet page sources individually. It seemed like a better idea to do it in a way that was independent of the underlying PageSource.

findepi replied:
We're constructing this for ORC and Parquet separately, right?

- would it be possible to use ORC `ColumnAdaptations` for ORC only?
- maybe later we add similar concept to Parquet?

thoughts?

alexjo2144 replied:
Sure, we could do that for ORC and have a follow up to add ColumnAdaptation support to Parquet and eventually Avro

alexjo2144 committed:
Move populating Iceberg partition values out of IcebergPageSource

Includes a Factory which skips the reconstruction of the Page if
there are no constant values to populate.

alexjo2144 committed:
Add ORC row position adaptation

alexjo2144 committed:
Iceberg support for reading v2 row level deletes

Co-authored-by: Jack Ye <yzhaoqin@amazon.com>
Co-authored-by: Konrad Dziedzic <konraddziedzic@gmail.com>

alexjo2144 commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/delete/DummyFileScanTask.java:
```diff
@@ -0,0 +1,231 @@
+import static java.util.Objects.requireNonNull;
+
+// TODO: This wrapper is necessary until the constructors of the Iceberg DeleteFilter are made more specific
+// Remove after upgrading to Iceberg with https://github.com/apache/iceberg/pull/4381
```
Yeah, the spot where this is used will not compile when the Iceberg dependency is updated.

alexjo2144 commented:
Updated based on comments, thanks for the reviews.

Red build is unrelated (kudu and ldap)

findinpath commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/ConstantPopulatingPageSource.java:
```diff
@@ -0,0 +1,211 @@
+        implements ConnectorPageSource
+{
+    private final ConnectorPageSource delegate;
+    private final Block[] constantColumns;
```
We're blocking here N `Block` instances and `N` `int` delegate indexes for `N` columns.
It is indeed rather straightforward to read, but I think we can do better.

Proposal: a rather naive idea which should work, but adds complexity in figuring out what is happening:
- constantColumns: an array of `B` `Block` corresponding to constant columns
- delegateColumnsIndexes: n array of `D` `int` delegate indexes 
- columnsIndexes: an array of `N` `int` containing the indexes in the block and delegate indexes array
- `B` + `D` = `N`

`columnsIndexes` have the following value structure:
- if pointing to a constant column index, the index value will be between `0` and `constantColumns.size()`
- if pointing to a delegate column index, the index value will be betwen `constantColumns.size()` and `N`

Depending on the value of the index, we know in which array to look for the value we want.

findinpath commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcPageSource.java:
```diff
@@ -398,7 +403,7 @@ public UpdatedRowAdaptation(HiveUpdateProcessor updateProcessor, List<HiveColumn
 
         @Override
-        public Block block(Page sourcePage, MaskDeletedRowsFunction maskDeletedRowsFunction, long filePosition)
+        public Block block(Page sourcePage, MaskDeletedRowsFunction maskDeletedRowsFunction, long filePosition, OptionalLong startRowId)
```
Apart from `PositionAdaptation` where it is only checked for being empty, I don't see `startRowId` actually being used anywhere. Is this intentional?

alexjo2144 replied:
It's not, but I'd rather leave it in even if just for that isEmpty check. If that Adapter is used in a situation where the value is present it'll silently produce incorrect results.

findinpath commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergTableHandle.java:
```diff
@@ -56,6 +57,7 @@ public IcebergTableHandle(
             @JsonProperty("tableName") String tableName,
             @JsonProperty("tableType") TableType tableType,
             @JsonProperty("snapshotId") Optional<Long> snapshotId,
+            @JsonProperty("schemaJson") String schemaJson,
```
schemaJson -> icebergSchema ?

We already have `schemaName` field in this class which makes this new field rather ambiguous.

findepi replied:
then `tabelSchemaJson` (adding "iceberg" generally doesn't mean anything within iceberg connector)

findinpath commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSource.java:
```diff
@@ -72,7 +100,23 @@ public Page getNextPage()
+
+            int positionCount = dataPage.getPositionCount();
+            int[] positionsToKeep = new int[positionCount];
+            try (CloseableIterable<TrinoRow> filteredRows = deleteFilter.filter(CloseableIterable.withNoopClose(TrinoRow.fromPage(columnTypes, dataPage, positionCount)))) {
```
```
            try (CloseableIterable<TrinoRow> rows = CloseableIterable.withNoopClose(TrinoRow.fromPage(columnTypes, dataPage, positionCount));
                    CloseableIterable<TrinoRow> filteredRows = deleteFilter.filter(rows)) {
```

findepi replied:
there is no point to ensure to close `TrinoRow.fromPage`, since it's not a resource.

findepi commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergV2.java:
```diff
@@ -134,7 +134,7 @@ public void testV2TableWithPositionDelete()
 
         icebergTable.newRowDelta().addDeletes(writer.toDeleteFile()).commit();
-        assertQueryFails("SELECT * FROM " + tableName, "Iceberg tables with delete files are not supported: tpch." + tableName);
+        assertQuery("SELECT count(*) FROM " + tableName, "VALUES 24");
```
This is a test of Trino internals vs Iceberg API.
Spark compatibility is covered in a product tests. Both this tests should exist.

We can _also_ add a test with static resources. 
We probably should do this also for v1 tables, as they have different metadata entries, etc. @findinpath can you please file an issue?

findinpath replied:
https://github.com/trinodb/trino/issues/11828

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/delete/TrinoDeleteFile.java:
```diff
@@ -0,0 +1,341 @@
+
+    @JsonProperty("lowerBounds")
+    @Nullable
+    public Map<Integer, byte[]> lowerBoundsAsByteArray()
```
for seralization; i don't think it requires a comment.

findepi commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergSparkCompatibility.java:
```diff
@@ -1412,6 +1403,26 @@ public void testStatsAfterAddingPartitionField()
+                "TBLPROPERTIES ('format-version'='2', 'write.delete.mode'='merge-on-read'," +
+                "'write.format.default'='" + tableStorageFormat.name() + "'," +
+                "'write.delete.format.default'='" + deleteFileStorageFormat.name() + "')");
+        onSpark().executeQuery("INSERT INTO " + sparkTableName + " VALUES (1, 2), (2, 2), (3, 2), (11, 12), (12, 12), (13, 12)");
+        onSpark().executeQuery("CALL " + SPARK_CATALOG + ".system.rewrite_data_files(table=>'" + TEST_SCHEMA_NAME + "." + tableName + "', options => map('min-input-files','1'))");
```
Document why you do INSERT _and_ a rewrite.

findepi commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergSparkCompatibility.java:
```diff
@@ -1412,6 +1403,26 @@ public void testStatsAfterAddingPartitionField()
+        onSpark().executeQuery("INSERT INTO " + sparkTableName + " VALUES (1, 2), (2, 2), (3, 2), (11, 12), (12, 12), (13, 12)");
+        onSpark().executeQuery("CALL " + SPARK_CATALOG + ".system.rewrite_data_files(table=>'" + TEST_SCHEMA_NAME + "." + tableName + "', options => map('min-input-files','1'))");
+        onSpark().executeQuery("DELETE FROM " + sparkTableName + " WHERE a = 13");
+        onSpark().executeQuery("DELETE FROM " + sparkTableName + " WHERE b = 2");
```
Document why two deletes. Would one be enough?

alexjo2144 replied:
One deletes a row in a file, the other deletes a whole partition. Can add a comment

findepi commented:
CI https://github.com/trinodb/trino/issues/11016

alexjo2144 commented:
Applied comments and added more more test I meant to have which has dereference pushdown.

findepi commented:
PT failure looks related

```
tests               | 2022-04-07 00:42:45 INFO: FAILURE     /    io.trino.tests.product.iceberg.TestIcebergSparkCompatibility.testTrinoReadsSparkRowLevelDeletes [PARQUET, PARQUET] (Groups: profile_specific_tests, iceberg) took 4.1 seconds
```

zhaoyuxiaoxiao commented:
@groobyming Hello, How do you set the "copy on Write" mode of your Iceberg table？
